executor/common_kvm_arm64_syzos.h

```diff
@@ -185,3 +203,444 @@ GUEST_CODE static void guest_handle_hvc(struct api_call_smccc* cmd)
+	// Set up the vector table.
+	asm(R"(
+		adr x1, guest_vector_table
+		msr vbar_el1, x1
+		msr daifclr, #0b1111
+	)"
+	    :
+	    :
+	    : "x1");
+}
+
+// Registers saved by one_irq_handler() and received by guest_irq_handler().
+struct ex_regs {
+	uint64 regs[31];
+	uint64 sp;
+	uint64 pc;
+	uint64 pstate;
+};
+
+// Placeholder function to declare one_irq_handler() inside the assembly blob. We cannot put it
+// into a separate .S file, because syzkaller requires a standalone header for reproducers.
+__attribute__((used))
+GUEST_CODE static void
+one_irq_handler_fn()
+{
+	asm volatile(
+	    R"(.global one_irq_handler
+	       one_irq_handler:
+	       # Allocate 34 * uint64 for struct ex_regs.
+	       add sp, sp, #-16 * 17
+	       # Store registers x0-x29 on the stack.
+	       stp x0, x1, [sp, #16 * 0]
+	       stp x2, x3, [sp, #16 * 1]
+	       stp x4, x5, [sp, #16 * 2]
+	       stp x6, x7, [sp, #16 * 3]
+	       stp x8, x9, [sp, #16 * 4]
+	       stp x10, x11, [sp, #16 * 5]
+	       stp x12, x13, [sp, #16 * 6]
+	       stp x14, x15, [sp, #16 * 7]
+	       stp x16, x17, [sp, #16 * 8]
+	       stp x18, x19, [sp, #16 * 9]
+	       stp x20, x21, [sp, #16 * 10]
+	       stp x22, x23, [sp, #16 * 11]
+	       stp x24, x25, [sp, #16 * 12]
+	       stp x26, x27, [sp, #16 * 13]
+	       stp x28, x29, [sp, #16 * 14]
+
+	       add x1, sp, #16 * 17
+	       # Store x30 and SP (before allocating ex_regs).
+	       stp x30, x1, [sp, #16 * 15] 
+
+	       # ELR_EL1 holds the PC to return to.
+	       mrs x1, elr_el1
+	       # SPSR_EL1 is the saved PSTATE.
+	       mrs x2, spsr_el1
+	       # Also store them to ex_regs.
+	       stp x1, x2, [sp, #16 * 16]
+
+	       # Call guest_irq_handler(ex_regs).
+	       mov x0, sp
+	       bl guest_irq_handler
+
+	       # Restore ELR_EL1 and SPSR_EL1.
+	       ldp x1, x2, [sp, #16 * 16]
+	       msr elr_el1, x1
+	       msr spsr_el1, x2
+
+	       # Restore the GP registers x0-x30 (ignoring SP).
+	       ldp x30, xzr, [sp, #16 * 15]
+	       ldp x28, x29, [sp, #16 * 14]
+	       ldp x26, x27, [sp, #16 * 13]
+	       ldp x24, x25, [sp, #16 * 12]
+	       ldp x22, x23, [sp, #16 * 11]
+	       ldp x20, x21, [sp, #16 * 10]
+	       ldp x18, x19, [sp, #16 * 9]
+	       ldp x16, x17, [sp, #16 * 8]
+	       ldp x14, x15, [sp, #16 * 7]
+	       ldp x12, x13, [sp, #16 * 6]
+	       ldp x10, x11, [sp, #16 * 5]
+	       ldp x8, x9, [sp, #16 * 4]
+	       ldp x6, x7, [sp, #16 * 3]
+	       ldp x4, x5, [sp, #16 * 2]
+	       ldp x2, x3, [sp, #16 * 1]
+	       ldp x0, x1, [sp, #16 * 0]
+
+	       add sp, sp, #16 * 17
+
+	       # Use ERET to exit from an exception.
+	       eret)"
+	    :
+	    :
+	    : "x0", "x1", "x2", "x3", "x4", "x5", "x6", "x7", "x8", "x9", "x10", "x11", "x12", "x13",
+	      "x14", "x15", "x16", "x17", "x18", "x19", "x20", "x21", "x22", "x23", "x24", "x25",
+	      "x26", "x27", "x28", "x29", "x30", "memory");
+}
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+__attribute__((used))
```

Comment:
What happens w/o __attribute__((used))? The symbol is used, does linker still remove it?  How is one_irq_handler_fn linked then?

Reply:
None of the functions marked as used are used from the C standpoint, so the compiler doesn't like them:

```
executor/common_kvm_arm64_syzos.h:612:24: error: 'void guest_vector_table_fn()' defined but not used [-Werror=unused-function]
  612 | GUEST_CODE static void guest_vector_table_fn()
```